### PR TITLE
Add MS Teams Notifier To Notification Hook

### LINKED ISSUE #318: ⚓️  Add a generic notification Hook with message templates
## New Hook implementation request

**Is your feature request related to a problem? Please describe.**
As _secureCodeBox user_ i would like to use a more generic `notification Hook` with specific and customizable message templates to implement my own notifications, based on **MS Teams** and **Slack**. 

**Describe the solution you'd like**

- This issue should combine the issues/prs #310 and #163 in a new hook with the name `notification-hook`.
- The _message templates_ for **MS Teams** and **Slack** must be implemented as independent mustache templates. There is an old implementation of the MS Teams hook which could be an example of this idea https://github.com/secureCodeBox/secureCodeBox/pull/163/commits/3881d7b7842a2b41c4efe24caaef6359e3873a3e
- The _message templates_ should be _referenced by name_ to configure them via HelmChart
- As a user i want to add my own mustache templates via configMap.
- As a user i want to configure multiple outgoing message endpoinds with different message templates

**Describe alternatives you've considered**

Implementing each messenger in a separate hook. But this would result in a lot of duplicate code.

### Example HelmChart Configuration (must have)

```yaml
notifications:
  - webhook:
        # notification.url -- The URL of your notification service endpoint (Teams, Slack, RocketChat,...)
        url: "http://msteams.example.com"
        # notification.template -- The name of the message template that should be used [messageCard | adaptiveCard].
        template: general-alert-msteams.mustache
        # notification.rules -- A optional rule definition that can be used to describe in wich case a notification must be fired. If not defined / empty each scan result will be notified.
        rules: []
          - matches:
               anyOf:
                 - severity: "high"
                 - severity: "medium"
    vulnerabilityManagement:
      # vulnerabilityManagement.enabled If enabled a direct link to your vulnerability management system will be added at the bottom of each message. 
      enabled: false
      # vulnerabilityManagement.name The name of your vulnerability management system wich you reference to.
      name: Kibana Dashboard
      # vulnerabilityManagement.url The URL to your vulnerability management system wich will be added at the bottom of each message.
      url: "https://your-kibana-service.url/your-dashboard-path"
      # vulnerabilityManagement.url A special URL used to deep link into your vulnerability management system. The placeholder {{uid}} will be replaced with the current scanId (uid).
      findingsUrl: "https://your-kibana-service.url/your-dashboard-path/filter:{{uid}}"
  - webhook:
        # notification.url -- The URL of your notification service endpoint (Teams, Slack, RocketChat,...)
        url: "http://slack.example.com"
        # notification.template -- The name of the message template that should be used [messageCard | adaptiveCard].
        template: team-warning-slack.mustache
        # notification.rules -- A optional rule definition that can be used to describe in wich case a notification must be fired. If not defined / empty each scan result will be notified.
        rules: []
          - matches:
               anyOf:
                 - location: "https://myteam.application.dev"
    vulnerabilityManagement:
      # vulnerabilityManagement.enabled If enabled a direct link to your vulnerability management system will be added at the bottom of each message. 
      enabled: true
      # vulnerabilityManagement.name The name of your vulnerability management system wich you reference to.
      name: Defect Dojo
      # vulnerabilityManagement.url The URL to your vulnerability management system wich will be added at the bottom of each message.
      url: "https://your-kibana-service.url/your-dashboard-path"
      # vulnerabilityManagement.url A special URL used to deep link into your vulnerability management system. The placeholder {{uid}} will be replaced with the current scanId (uid).
      findingsUrl: "https://your-defectdojo-service.url/my-team-path?finding={{uid}}"

```

### Mail Example HelmChart Configuration (optional)

```yaml
notifications:
  - mail:
        # notification.address -- The reciever mail address of this notification
        address: "myteam@example.com"
        # notification.subject -- The mail subject
        subject: "[SCB] You got Mail!"
        # notification.template -- The name of the message template that should be used [messageCard | adaptiveCard].
        template: general-email.mustache
        # notification.rules -- A optional rule definition that can be used to describe in wich case a notification must be fired. If not defined / empty each scan result will be notified.
        rules: []
          - matches:
               anyOf:
                 - severity: "high"
                 - severity: "medium"
    vulnerabilityManagement:
      # vulnerabilityManagement.enabled If enabled a direct link to your vulnerability management system will be added at the bottom of each message. 
      enabled: false
      # vulnerabilityManagement.name The name of your vulnerability management system wich you reference to.
      name: Kibana Dashboard
      # vulnerabilityManagement.url The URL to your vulnerability management system wich will be added at the bottom of each message.
      url: "https://your-kibana-service.url/your-dashboard-path"
      # vulnerabilityManagement.url A special URL used to deep link into your vulnerability management system. The placeholder {{uid}} will be replaced with the current scanId (uid).
      findingsUrl: "https://your-kibana-service.url/your-dashboard-path/filter:{{uid}}"
```

**Additional context**

- Template Engine: https://mustache.github.io/
- #310
- #163
- Message Examples: https://github.com/secureCodeBox/secureCodeBox/tree/main/hooks/teams-webhook/__testfiles__
- https://api.slack.com/messaging/webhooks
- https://docs.microsoft.com/en-us/microsoftteams/platform/webhooks-and-connectors/how-to/connectors-using

## Steps to implement a new Hook

- [x] Create a new folder with the name of the [hook here](https://github.com/secureCodeBox/secureCodeBox/tree/master/hooks)
- [x] Add a README and give a brief overview of the scanner and its configuration options.
- [ ] Add (optional) a Dockerfile for the scanner if there is no existing one publicly available on dockerHub
- [x] Use the [Hook-SDK](https://github.com/secureCodeBox/secureCodeBox/tree/master/hook-sdk) to implement a new hook (currently based on NodeJS)
- [ ] Add unit tests with at minimum 80% test coverage
- [ ] Add some example scan.yaml and finding.yaml files in the example folder
- [x] Implement a new integration test for the hook [here](https://github.com/secureCodeBox/secureCodeBox/tree/master/tests/integration)

### PR DESCRIPTION
**Pull request by fuhrmeistery**
- [Refactor] Create New AbstractWebHookNotifier
- Add MSTeamsNotifier
- Add Option To Render Nunjucks Templates
- Add Documentation For MSTeamsNotifier

closes #318

### CODE REVIEW

File: hooks/notification/notification-templates/msteams-messageCard.njk [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        weight: Bolder
+        text: "New {{ scan.spec.scanType }} security scan results are available!"
+        wrap: true
+      - type: ColumnSet
+        columns:
+        - type: Column
+          items:
+          - type: Image
+            style: Person
+            url: "https://securecodebox.io/favicon.png"
```
Comment (rseedorff) [inline hooks/notification/notification-templates/msteams-messageCard.njk:19]:
```suggestion
            url: "https://www.securecodebox.io/favicon.png"
```

File: hooks/notification/notification-templates/msteams-messageCard.njk [hunk 2]
```diff
@@ -0,0 +1,66 @@
+type: message
```
Comment (rseedorff) [inline hooks/notification/notification-templates/msteams-messageCard.njk:1]: Is the `hooks/notification/notification-templates/.msteams-messageCard.njk.swp` binary really necessary? Looks a bit strange... Maybe add the file extension *.swp to .gitignore?

File: hooks/notification/notification-templates/msteams-messageCard.njk [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        wrap: true
+        weight: Bolder
+        size: Medium
+      - type: FactSet
+        facts:
+          {% for severity, counter in scan.status.findings.severities -%}
+          - title: "{{ severity }}:"
+            value: "{{ counter }}"
+          {% endfor %}
+      {% if args["VULNMANAG_ENABLED"] == true -%}
```
Comment (rseedorff) [inline hooks/notification/notification-templates/msteams-messageCard.njk:56]: A documentation how to define the `VULNMANAG_ENABLED` parameter would be helpful. Is there a specific HelmChart Value that needs to be defined?

File: hooks/notification/notification-templates/msteams-messageCard.njk [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      actions:
+      - type: Action.OpenUrl
+        title: Open Dashboard
+        url: "{{ renderString(args["dashboardUrl"].toString(), { uid: scan.metadata.uid }) }}"
```
Comment (rseedorff) [inline hooks/notification/notification-templates/msteams-messageCard.njk:60]: A documentation how to define the `dashboardUrl ` parameter would be helpful

File: hooks/notification/notification-templates/msteams-messageCard.njk [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      - type: Action.OpenUrl
+        title: Show Results in Dashboard
+        url: "{{ renderString(args["dashboardFindingsUrl"], { uid: scan.metadata.uid }) }}"
```
Comment (rseedorff) [inline hooks/notification/notification-templates/msteams-messageCard.njk:63]: A documentation how to define the `dashboardFindingsUrl ` parameter would be helpful

**Review by rseedorff — CHANGES_REQUESTED** [review]

File: hooks/notification/notification-templates/msteams-messageCard.njk (hunk 2 above)
Comment (fuhrmeistery) [reply hooks/notification/notification-templates/msteams-messageCard.njk:1]: Nope this was not necessary. Good catch :)

File: hooks/notification/README.md.gotmpl [hunk 6]
```diff
@@ -210,6 +211,37 @@ env:
...
+notificationChannels:
+  - name: ms-teams
+    type: ms-teams
+    template: msteams-messageCard
+    rules: []
+    endPoint: "https://somewhere.xyz/sadf12"
+env:
+  - name: VULNMANAG_ENABLED
+    value: true
+  - name: dashboardUrl
```
Comment (rseedorff) [inline hooks/notification/README.md.gotmpl:239]:
```suggestion
  - name: VULNMANAG_DASHBOARD_URL
```
Env Vars should be UpperCase to be more consistent.

File: hooks/notification/README.md.gotmpl [hunk 7]
```diff
@@ -210,6 +211,37 @@ env:
...
+    value: "somedashboard.url"
+  - name: dashboardFindingsUrl
```
Comment (rseedorff) [inline hooks/notification/README.md.gotmpl:241]:
```suggestion
  - name: VULNMANAG_DASHBOARD_FINDINGS_URL
```

File: hooks/notification/README.md.gotmpl [hunk 8]
```diff
@@ -210,6 +211,37 @@ env:
...
 ```
 
+### Configuration Of A MS Teams Notification
+
+To configure a MS Teams notification you need to set the type to `ms-teams`.
+In `endPoint` you need to specify the MS Teams webhook.
+To use the template provided by the secureCodeBox set template to `msteams-messageCard`.
+
+The default template allows you to specify an additional set of information.
+If you you use the persistence-elastic or use some kind of dashboard, you can set the variable `VULNMANAG_ENABLED` to true, `dashboardUrl` to the URL of your dashboard.
```
Comment (rseedorff) [inline hooks/notification/README.md.gotmpl:221]:
```suggestion
If you you use an external web based vulnerability management system with some kind of dashboard, you can set the variable `VULNMANAG_ENABLED` to true and point the `VULNMANAG_DASHBOARD_URL` to the URL of your vulnerability management dashboard.
```

File: hooks/notification/README.md.gotmpl [hunk 9]
```diff
@@ -210,6 +211,37 @@ env:
...
+This will add a button in the notification that opens your dashboard.
```
Comment (rseedorff) [inline hooks/notification/README.md.gotmpl:222]:
```suggestion
This will add a button in the notification that links directly to your dashboard.
```

File: hooks/notification/README.md.gotmpl [hunk 10]
```diff
@@ -210,6 +211,37 @@ env:
...
+You can also add a button that opens your findings directly in your dashboard.
+To do this you need to specify `dashboardFingingsUrl`.
```
Comment (rseedorff) [inline hooks/notification/README.md.gotmpl:224]:
```suggestion
To do this you need to specify `VULNMANAG_DASHBOARD_FINDINGS_URL`.
```

File: hooks/notification/notification-templates/msteams-messageCard.njk [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
```
Comment (rseedorff) [inline hooks/notification/notification-templates/msteams-messageCard.njk:63]:
```suggestion
        url: "{{ renderString(args["VULNMANAG_DASHBOARD_FINDINGS_URL"], { uid: scan.metadata.uid }) }}"
```

File: hooks/notification/notification-templates/msteams-messageCard.njk [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        facts:
+          {% for severity, counter in scan.status.findings.severities -%}
+          - title: "{{ severity }}:"
+            value: "{{ counter }}"
+          {% endfor %}
+      {% if args["VULNMANAG_ENABLED"] == true -%}
+      actions:
+      - type: Action.OpenUrl
+        title: Open Dashboard
+        url: "{{ renderString(args["dashboardUrl"].toString(), { uid: scan.metadata.uid }) }}"
```
Comment (rseedorff) [inline hooks/notification/notification-templates/msteams-messageCard.njk:60]:
```suggestion
        url: "{{ renderString(args["VULNMANAG_DASHBOARD_URL"].toString(), { uid: scan.metadata.uid }) }}"
```

**Review by rseedorff — CHANGES_REQUESTED** [review]

**Review by J12934 — APPROVED** [review]
